hibernate-core/src/test/java/org/hibernate/test/stream/basic/JpaStreamTest.java

```diff
@@ -70,12 +79,100 @@ public void testQueryStream() {
 		} );
 	}
 
+	@Test
+	@TestForIssue( jiraKey = "HHH-13872")
+	@RequiresDialect(H2Dialect.class)
+	public void testStreamCloseOnTerminalOperation() {
+		doInHibernate( this::sessionFactory, session -> {
+			session.createQuery( "delete from MyEntity" ).executeUpdate();
+
+			for ( int i = 1; i <= 10; i++ ) {
+				MyEntity e = new MyEntity();
+				e.id = i;
+				e.name = "Test";
+				session.persist( e );
+			}
+		} );
+
+		doInHibernate( this::sessionFactory, session -> {
+			Stream<MyEntity> stream = session
+					.createQuery( "SELECT me FROM MyEntity me" )
+					.getResultStream();
+
+			ResourceRegistry resourceRegistry = resourceRegistry(session);
+			assertTrue( resourceRegistry.hasRegisteredResources() );
+
+			List<MyEntity> entities = stream.collect( Collectors.toList() ) ;
+			assertEquals(10, entities.size());
+
+			assertFalse( resourceRegistry.hasRegisteredResources() );
+		} );
+
+		doInHibernate( this::sessionFactory, session -> {
+			Stream<MyEntity> stream = session
+					.createQuery( "SELECT me FROM MyEntity me" )
+					.getResultStream();
+
+			ResourceRegistry resourceRegistry = resourceRegistry(session);
+			assertTrue( resourceRegistry.hasRegisteredResources() );
+
+			int sum = stream.mapToInt( MyEntity::getId ).sum();
+			assertEquals(55, sum);
+
+			assertFalse( resourceRegistry.hasRegisteredResources() );
+		} );
+
+		doInHibernate( this::sessionFactory, session -> {
+			Stream<MyEntity> stream = session
+					.createQuery( "SELECT me FROM MyEntity me" )
+					.getResultStream();
+
+			ResourceRegistry resourceRegistry = resourceRegistry(session);
+			assertTrue( resourceRegistry.hasRegisteredResources() );
+
+			long result = stream.mapToLong( entity -> entity.id * 10 ).min().getAsLong();
+			assertEquals(10, result);
+
+			assertFalse( resourceRegistry.hasRegisteredResources() );
+		} );
+
+		doInHibernate( this::sessionFactory, session -> {
+			Stream<MyEntity> stream = session
+					.createQuery( "SELECT me FROM MyEntity me" )
+					.getResultStream();
+
+			ResourceRegistry resourceRegistry = resourceRegistry(session);
+			assertTrue( resourceRegistry.hasRegisteredResources() );
+
+			double result = stream.mapToDouble( entity -> entity.id * 0.1D ).max().getAsDouble();
+			assertEquals(1, result, 0.1);
+
+			assertFalse( resourceRegistry.hasRegisteredResources() );
+		} );
+	}
+
+	private ResourceRegistry resourceRegistry(Session session) {
+		SharedSessionContractImplementor sharedSessionContractImplementor = (SharedSessionContractImplementor) session;
+		JdbcCoordinator jdbcCoordinator = sharedSessionContractImplementor.getJdbcCoordinator();
+		return jdbcCoordinator.getLogicalConnection().getResourceRegistry();
+	}
+
```

Comment:
One good edge case for this PR is to verify the following scenario:

- some stream client code has closed stream already (to adhere to best practice, maybe via 'try' block);
- this new feature was added without client knowing and the stream might be closed twice;

It would be excellent to add a testing case to ensure the above common scenario won't throw some exception (not only for now, but for the long future).

Reply:
Yes, we should totally add that test case. Luckily, the `AsbtractScrollableResults` returns immediately from `close` if the RS was already closed:

```
if ( this.closed ) {
	// noop if already closed
	return;
}
```

Reply:
Yeah, I noticed that, but adding testing case would make us sleep well if the above code was refactored.